File: generated_code/cpp/solution_2299_3.cpp
#include <vector>
#include <algorithm>
#include <iostream>

// Definition for singly-linked list.
// Provided in problem statement; explicitly defined here to ensure self-contained, compilable code.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        // Algorithm: Two Pointers / In-place Modification
        // Time Complexity: O(N) where N is the number of nodes. We traverse the list exactly once.
        // Space Complexity: O(1) auxiliary space as we modify the existing list in-place.
        
        ListNode* modify = head; // Pointer to overwrite values and rebuild the sequence
        ListNode* curr = head->next; // Pointer to traverse the original linked list
        
        int current_sum = 0;
        
        while (curr != nullptr) {
            if (curr->val == 0) {
                // We reached a 0 separator. Update the value of the modify node,
                // and advance the modify pointer to its next available position.
                modify->val = current_sum;
                
                // If this 0 is the absolute end of the list, we decouple the remaining nodes.
                if (curr->next == nullptr) {
                    modify->next = nullptr;
                } else {
                    modify = modify->next;
                }
                
                // Reset sum for the next segment
                current_sum = 0;
            } else {
                // Accumulate the segment sum
                current_sum += curr->val;
            }
            curr = curr->next;
        }
        
        return head;
    }
};

File: generated_code/python/solution_2299_3.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next

class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Two-pointer in-place approach.
        # Time Complexity: O(N) where N is the number of nodes in the linked list.
        # Space Complexity: O(1) as we modify the existing list nodes in-place.
        
        # 'modify' will track the node where we overwrite the summed values.
        # 'curr' will traverse the list to calculate the sums.
        modify = head.next
        curr = modify
        
        while curr is not None:
            current_sum = 0
            # Accumulate values until we encounter the next 0 node.
            while curr and curr.val != 0:
                current_sum += curr.val
                curr = curr.next
            
            # Overwrite the value of the modify node with the accumulated sum.
            modify.val = current_sum
            
            # Move curr to the node after the 0 node.
            curr = curr.next
            
            # Link the current modified node to the next segment's start node.
            modify.next = curr
            
            # Move the modify pointer forward for the next segment.
            modify = modify.next
            
        return head.next